docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst

```diff
@@ -0,0 +1,1946 @@
+
+Finally, you can migrate your live data from your previous dars to the
+new LF1.16 dars, using one of the existing downtime upgrade techniques
+listed in `What are Zero Downtime Smart Contract Upgrades <#what-are-zero-downtime-smart-contract-upgrades>`__.
+
+The Upgrade Model in Depth - Reference
+--------------------------------------
+
+You can find the in-depth upgrading model, which can be used as a reference
+for valid upgrades, :ref:`here <upgrade_model_reference>`.
+
+Components
+==========
+
+Ledger API
+----------
+
+Until the introduction of SCU, templates in requests on the Ledger API
+could only be referenced by the template-id, with the template fully
+qualified name of format ``<package-id>:<module-name>:<template-name>``.
+
+With SCU, we introduce a more generic template reference of the format
+``#<package-name>:<module-name>:<template-name>``. This format is only a
+Ledger API concept and is meant to suggest to the Ledger API to perform
+a dynamic runtime resolution of packages in the Daml engine when
+generating the Daml transaction after command interpretation. This
+dynamic resolution is based on the existing upgradable (LF >= 1.16)
+package-ids pertaining to a specific ``package-name`` and is possible on the
+write path (command submission) and read path (Ledger API queries) as
+presented below.
+
+Dynamic package resolution in command submission
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Dynamic package resolution can happen in two cases during command
+submission:
+
+-  For command submissions that use the package-name selector
+   (``#<package-name>``) in the command’s templateId field (e.g. in a
+   create command :ref:`here <com.daml.ledger.api.v1.CreateCommand>`)
+
+-  For command submissions leading to Daml transactions that contain
+   actions exercised on interfaces. In this situation there may be
+   many versions of a template that implement the interface being
+   exercised.
+
+In these situations the following rules are followed to resolve the
+package-name to a package-id:
+
+-  By default, the participant resolves a package-name to the package-id
+   pertaining to the highest package version uploaded
+
+-  The command submitter can override the above-mentioned default
+   participant resolution by pinning package-ids in the Command’s
+   :ref:`package_id_selection_preference <com.daml.ledger.api.v1.Commands.package_id_selection_preference>`.
+   More specifically, this field is a list of package-ids that must
+   be explicitly used when resolving package-name *ambiguities* in
+   either command template-id or interface resolution.
+
+   -  **Note:** The Command’s
+      :ref:`package_id_selection_preference <com.daml.ledger.api.v1.Commands.package_id_selection_preference>`
+      must not lead to ambiguous resolutions for package-names,
+      meaning that it must not contain two package-ids pertaining to
+      the same package-name otherwise the submission will fail with
```

Comment:
```suggestion
      meaning that it must not contain two package-ids pertaining to
      the same package-name otherwise the submission will fail with
```
```suggestion
      meaning that it must not contain two package-ids pointing to
      packages with the same package-name, as otherwise the submission will fail with
```